ggshield/scan/scan_context.py

```diff
@@ -1,12 +1,35 @@
+import pathlib
+import platform
+import sys
 import uuid
 from dataclasses import dataclass
-from typing import Dict, Optional, Union
+from typing import Dict, Optional, Tuple, Union
 
 from ggshield import __version__
 
 from .scan_mode import ScanMode
 
 
+def get_os_info() -> Tuple[str, str]:
+    if sys.platform.lower() == "linux":
+        return parse_os_release()
+    else:
+        return platform.system().lower(), platform.version()
+
+
+def parse_os_release() -> Tuple[str, str]:
+    """Return both Linux OS-name and OS-Version"""
+    os_release_path = pathlib.Path("/etc/os-release")
+    with open(os_release_path) as f:
+        lines = f.readlines()
+
+    # Build a dictionary from the metadata
+    data_dict = {
+        key.strip(): value.strip() for key, value in (line.split("=") for line in lines)
```

Comment:
There is an issue here: `value` may be surrounded with double-quotes. These quotes must be stripped:

```
In [1]: from ggshield.scan.scan_context import parse_os_release

In [2]: parse_os_release()
Out[2]: ('ubuntu', '"22.04"')
```

The function must also be hardened: if it fails, no scans happen :/ Right now it is going to fail if:

- the file does not exist
- the file cannot be read
- the file contains a line without a `=` ([os-release man page](https://www.linux.org/docs/man5/os-release.html) says line starting with `#` are comments)
- one of ID or VERSION_ID keys are not present. The man page says VERSION_ID is optional. It is the case on Arch linux because it's a rolling-release distribution:
  ```
  $ docker run --rm archlinux:latest cat /etc/os-release
  NAME="Arch Linux"
  PRETTY_NAME="Arch Linux"
  ID=arch
  BUILD_ID=rolling
  ANSI_COLOR="38;2;23;147;209"
  HOME_URL="https://archlinux.org/"
  DOCUMENTATION_URL="https://wiki.archlinux.org/"
  SUPPORT_URL="https://bbs.archlinux.org/"
  BUG_REPORT_URL="https://bugs.archlinux.org/"
  LOGO=archlinux-logo
  ```

If the file cannot be read, we should fallback to `linux` for the name and `unknown` for the version.

If VERSION_ID is not set it should also be `unknown`.


I suggest you add a test just for parse_os_release(). To make it easier to test, you can modify it to take the path of the file to parse instead, this way you can pass it any file, using the `tmp_path` Pytest fixture. Then the code would look like this:

```
def get_os_info() -> Tuple[str, str]:
    if sys.platform.lower() == "linux":
        return parse_os_release(Path("/etc/os-release"))
    else:
        return platform.system().lower(), platform.version()


def parse_os_release(os_release_path: Path) -> Tuple[str, str]:
    ...
```

Reply:
One more thing: you can add `@lru_cache(None)` (from functools module) to `get_os_info()` so that the results are cached.

Reply:
Thanks @agateau-gg. I learned many new things from this PR :pray: 
By the way, I am skipping the new tests when the platform != Linux (it will not be relevant otherwise). 

This is the decorator:
```python
@pytest.mark.skipif(
    sys.platform.lower() != "linux", reason="This test is only relevant on Linux."
)
```

Let me know what you think about the state of this PR :smile_cat: